### PR TITLE
Add `CCCL_DISABLE_NVTX` macro

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes: #2172

### CODE REVIEW

File: cub/test/test_nvtx_disabled.cu [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#define CUB_DETAIL_BEFORE_NVTX_RANGE_SCOPE(name) static_assert(false, "");
+#define CCCL_DISABLE_NVTX
+
+#include <cub/device/device_for.cuh>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cuda/std/functional>
+
```
Comment (bernhardmgruber) [inline cub/test/test_nvtx_disabled.cu:9]: TODO: Maybe add this suggestion from @gevtushenko:
```
#ifdef CCCL_DISABLE_NVTX
#ifdef NVTX_VERSION
#error "Should not include NVTX when CCCL_DISABLE_NVTX is defined"
#endif
#endif
```
To make sure we are also not including the NVTX headers by accident.

**Comment by bdice** [comment]
I would propose matching what RMM does: `option(RMM_NVTX "Build RMM with NVTX support" OFF)`

- Name it `CCCL_NVTX` in CMake and in the macro
- Default to OFF in CMake and in the macro usage
  - If including headers directly and not using CMake to find CCCL, NVTX should be off

**Review by miscco — APPROVED** [review]

**Comment by bernhardmgruber** [comment]
> I would propose matching what RMM does: `option(RMM_NVTX "Build RMM with NVTX support" OFF)`

The issue with an enable macro is that we want NVTX ranges to be on by default. I measured the overhead and it's negligible. So if you compile without cmake and define no macros, it should be on. Thus I went for a disable macro. This also mirrors what NVTX itself does.

@gevtushenko, since you supervised me during the NVTX range integration, would you rather prefer an enable macro with NVTX off by default?

> RAPIDS uses CCCL heavily and if we get NVTX ranges from CCCL there will be so many ranges in the profile that we can't parse it.

You probably already have NVTX ranges in CCCL for some time :) We introduced them at the end of April 2024 in #1657. In any case, you can disable them now at your liking.

> I'm not sure if you need special logic to handle the global `NVTX_DISABLE` definition. I think that is handled by NVTX itself, but please check me on that.

We wouldn't need additional logic, but it was also trivial to remove our own scope guards, so we went for it.

File: cub/test/test_nvtx_disabled.cu (hunk 1 above)
Comment (bernhardmgruber) [reply cub/test/test_nvtx_disabled.cu:9]: Added.

**Comment by bernhardmgruber** [comment]
> @gevtushenko, since you supervised me during the NVTX range integration, would you rather prefer an enable macro with NVTX off by default?

@gevtushenko Ping.

File: cub/cub/detail/nvtx.cuh [hunk 2]
```diff
@@ -39,9 +39,9 @@
 
 // Enable the functionality of this header if:
 // * The NVTX3 C API is available in CTK
-// * NVTX is not explicitly disabled
+// * NVTX is not explicitly disabled (via CCCL_DISABLE_NVTX or NVTX_DISABLE)
 // * C++14 is availabl for cuda::std::optional
-#if __has_include(<nvtx3/nvToolsExt.h> ) && !defined(NVTX_DISABLE) && _CCCL_STD_VER >= 2014
+#if __has_include(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) && _CCCL_STD_VER >= 2014
```
Comment (gevtushenko) [inline cub/cub/detail/nvtx.cuh:44]: suggestion: document API macro as follows:

```cpp
#ifdef DOXYGEN_SHOULD_SKIP_THIS // Only parse this during doxygen passes:

/**
 * @def CCCL_DISABLE_NVTX
 *
 * Disables NVTX ...
 */
#  define CCCL_DISABLE_NVTX
...
```

**Review by gevtushenko — APPROVED** [review]
I agree with keeping NVTX on by default. Regarding the CMake option, we expose a lot of control on macro level rather than CMake because CCCL is occasionally used as header-only library with just `-Icccl/{thrust,cub,libcudacxx/include`. We can investigate a quality-of-life improvement that'd expose these control knobs as CMake options independently of this PR.

File: cub/cub/detail/nvtx.cuh (hunk 2 above)
Comment (bernhardmgruber) [reply cub/cub/detail/nvtx.cuh:49]: Great idea, added!

**Pull request merged**